src/brevitas/graph/equalize.py

```diff
@@ -482,73 +640,206 @@ def apply(self,
+                    # act_name = region.acts[0]
+                    act_module = name_to_module[act_name]
+                    hook_fn = partial(self.forward_stats_hook, name=act_name)
+                    self.hooks.append(act_module.register_forward_hook(hook_fn))
+
+        self.regions = [x for x in self.regions if x not in regions_to_drop]
+
+    def apply(self, alpha):
+        scale_factors = []
+        name_to_module = dict_name_to_module(self.graph_model, self.regions)
+        for region in self.regions:
+            act_module = [name_to_module[act_name] for act_name in region.acts]
+            list_of_act_val = [self.float_act_map[act_name] for act_name in region.acts]
+            sinks = [name_to_module[sink] for sink in region.sinks]
+            # Filter out scale_varying activations from the srcs
+            srcs = [
+                name_to_module[src]
+                for src in region.srcs
+                if not isinstance(name_to_module[src], _scale_varying_activations)]
+
+            list_of_insert_mul_node_fn = None
+            if self.add_mul_node and any([
+                    isinstance(act, _scale_varying_activations) for act in act_module]):
+                # Even though we iterate, this list will always have a single element by definition
+                list_of_insert_mul_node_fn = []
+                for act_name in region.acts:
+                    act_node = get_node(self.graph_model, act_name)
+                    list_of_insert_mul_node_fn.append(
+                        partial(self.insert_mul_node, act_node=act_node))
+            scale_factors.append(
+                _cross_layer_equalization(
+                    srcs,
+                    sinks,
+                    False,
+                    scale_computation_type=self.scale_computation_type,
+                    list_of_act_val=list_of_act_val,
+                    list_of_insert_mul_node_fn=list_of_insert_mul_node_fn,
+                    alpha=alpha))
+
+        return scale_factors
+
+    def remove_hooks(self):
+        for hook in self.hooks:
             hook.remove()
-        self.collect_inp_shape_hooks = []
 
-    def apply(self, graph_model: GraphModule, inp):
-        self.collect_inp_shapes(graph_model, inp)
-        groups = self.extract_groups(graph_model)
-        self.absorb_biases(groups)
-        return graph_model
+    def forward_stats_hook(self, module, inp, out, name):
+        # Compute stats along the batch dimension
+        if name not in self.float_act_map:
+            self.float_act_map[name] = self.scale_fn(out, dim=0)
+        else:
+            batch_data = torch.cat([self.float_act_map[name].unsqueeze(0), out], dim=0)
+            self.float_act_map[name] = self.scale_fn(batch_data, dim=0)
+
+    def insert_mul_node(self, scale, shape, axis, act_node):
+        broadcastable_shape = [1] * len(shape)
+        broadcastable_shape[axis] = shape[axis]
+        # Add Batch Dim
```

Comment:
I think we need more work around determining the batch size, e.g. for whenever batch_first=False in attention-y implementation.